### ISSUE TITLE
Ability to allow shared object inheritance

### ISSUE THREAD

**Issue by johanneshiry** [carueda/tscfg#64]
After their introduction, we found shared objects very useful in our configuration setup. Based on our experience working with shared objects, we would highly appreciate the availability of some kind of basic inheritance for shared object definition. Especially from the scala config perspective, this would reduce code duplicates when working with several shared object definitions that have one or more fields in common because one could pattern match against them. 

I have not thought about all details yet, because I'm unsure if this a feature that could be of interested to be included, but my first idea was something like:

```
#@define_pa
BaseModelConfig {
  uuids: [string]         
  scaling: double  
}

#@define
LoadModelConfig ex BaseModelConfig{
  modelBehaviour: string  
  reference: string       
}

#@define
FixedFeedInModelConfig ex BaseModelConfig
```

This would end up in either a sealed trait or an abstract class (whatever is easier) in the generated code (only thought about scala for now, maybe slightly different concept needs to be developed for java code):

```scala
  sealed trait BaseModelConfig {
    val scaling: scala.Double
    val uuids: scala.List[java.lang.String]
  }

  final case class LoadModelConfig(
                              scaling: Double,
                              uuids: List[String],
                              modelBehaviour: java.lang.String,
                              reference: java.lang.String
                            ) extends BaseModelConfig

  object LoadModelConfig {
    def apply(
               c: com.typesafe.config.Config,
               parentPath: java.lang.String,
               $tsCfgValidator: $TsCfgValidator
             ): SimonaConfig.LoadModelConfig = {
      SimonaConfig.LoadModelConfig(
        scaling = $_reqDbl(parentPath, c, "scaling", $tsCfgValidator),
        uuids = $_L$_str(c.getList("uuids"), parentPath, $tsCfgValidator),
        modelBehaviour =
          $_reqStr(parentPath, c, "modelBehaviour", $tsCfgValidator),
        reference = $_reqStr(parentPath, c, "reference", $tsCfgValidator)
      )
    }

    private def $_reqStr(
                          parentPath: java.lang.String,
                          c: com.typesafe.config.Config,
                          path: java.lang.String,
                          $tsCfgValidator: $TsCfgValidator
                        ): java.lang.String = {
      if (c == null) null
      else
        try c.getString(path)
        catch {
          case e: com.typesafe.config.ConfigException =>
            $tsCfgValidator.addBadPath(parentPath + path, e)
            null
        }
    }

    private def $_reqDbl(
                          parentPath: java.lang.String,
                          c: com.typesafe.config.Config,
                          path: java.lang.String,
                          $tsCfgValidator: $TsCfgValidator
                        ): scala.Double = {
      if (c == null) 0
      else
        try c.getDouble(path)
        catch {
          case e: com.typesafe.config.ConfigException =>
            $tsCfgValidator.addBadPath(parentPath + path, e)
            0
        }
    }
  }

final case class FixedFeedInModelConfig(scaling: Double, uuids: List[String]) extends BaseModelConfig

  object FixedFeedInModelConfig {
    def apply(
               c: com.typesafe.config.Config,
               parentPath: java.lang.String,
               $tsCfgValidator: $TsCfgValidator
             ): SimonaConfig.FixedFeedInModelConfig = {
      SimonaConfig.FixedFeedInModelConfig(
        scaling = $_reqDbl(parentPath, c, "scaling", $tsCfgValidator),
        uuids = $_L$_str(c.getList("uuids"), parentPath, $tsCfgValidator),
      )
    }

    private def $_reqDbl(
                          parentPath: java.lang.String,
                          c: com.typesafe.config.Config,
                          path: java.lang.String,
                          $tsCfgValidator: $TsCfgValidator
                        ): scala.Double = {
      if (c == null) 0
      else
        try c.getDouble(path)
        catch {
          case e: com.typesafe.config.ConfigException =>
            $tsCfgValidator.addBadPath(parentPath + path, e)
            0
        }
    }
  }

```

This is a first draft of the concept for now but I hope the general idea becomes clear. Looking forward to read what you think about it.

Cheers,

Johannes 

**tl;dr**
- inheritance for shared objects 
- added `final` modifier to scala case classes for scala best practices 🤓

**Comment by carueda** [comment]
Thanks for the suggestion!

Due to time constraints, here's at least a quick reaction:  Because of the [use of Typesafe Config itself for the parsing](https://github.com/carueda/tscfg/issues/51), the indicated syntax won't be possible exactly as given, but it might be something like this:

```
#@define
BaseModelConfig {
  uuids: [string]         
  scaling: double  
}

#@define extending BaseModelConfig
LoadModelConfig {
  modelBehaviour: string  
  reference: string       
}

#@define extending BaseModelConfig
FixedFeedInModelConfig ex BaseModelConfig
```

(btw,  #62 should probably be considered as part of any plans to address this)

**Comment by johanneshiry** [comment]
Thanks for the fast reply. For the syntax im totally fine if it needs to be adapted slightly as you proposed. The main intention behind my first draft was to keep it as short as possible but the required syntax might even be more expressive than the shorthand syntax. 

I consider preparing a first draft implementation and hand in a PR. Please let me know if there are any more suggestions additional to #62 that one might address in this PR too.

**Comment by johanneshiry** [comment]
@carueda I just finished my PR #65 with the new feature. I'm not adressing #62 for now as I'm unsure about the exact requirements with the enums.

Furthermore, based on my experience with the code so far, adding more complex structures might require some more extensive changes of the main code to avoid increasing cognitive complexity of some methods too much. But this is a very subjective view of a Scala novice.

**Comment by carueda** [comment]
@johanneshiry  Thanks much for your work on this!  Please bear with me while I find some time, hopefully within days, to review (other commitments are demanding all my time these days).

**Comment by johanneshiry** [comment]
no worries at all - there is no rush on this. :-)

**Comment by carueda** [comment]
@johanneshiry Thanks again.  I'll be preparing a new release while doing some further revision, probably also entering some observations in the tracker.
